WordPress/gutenberg
# Block Editor: Pressing enter at start of the text block prepends a block

## What?


### Before

Previously, you would have to use `Insert before` feature to prepend a block. Using the keyboard to achieve the same goal with the Paragraph and Heading blocks was only possible. You could do it by pressing `enter` with the caret at the start of the text block that contains content.

https://user-images.githubusercontent.com/699132/179499983-71784389-fd2d-4c5f-a32d-0fdf9f7b1ac5.mov

### After

This PR proposes bringing the same functionality that exists for the Paragraph and Heading blocks to other text blocks:

- Code
- Preformatted
- Read More
- Site Tagline
- Site Title
- Verse

https://user-images.githubusercontent.com/699132/179212805-345b4077-8c45-46e8-a075-7312f0690fe7.mov

It might be good to check if we can add it as a follow-up to other blocks like:

- Quote – the most recent version uses nested blocks, so it's a more complex task
- Pullquote – it's similar to the old Quote block, but the question is whether this block shouldn't follow the refactoring done to use nested blocks

I'm keeping this PR scoped because the aforementioned blocks have more complex usage or `RichText` component.

Another potential follow-up task is to improve the behavior when entering multiple new lines at the end of the block. Currently, you also get trapped inside text blocks that allow new lines in the content presented in the "before" video.

## Why?


To improve the writing flow experience and have more natural and consistent interaction when pressing enter when the caret is placed at the start of the text blocks.

## How?


The initial proposal mirrors the existing unstable method `__unstableOnSplitAtEnd` that inserts the block when the `enter` key is pressed at the end of the block. This PR proposes adding `__unstableOnSplitAtStart` function callback to `RichText` that gets called in the following conditions:
- there is any content in the contenteditable controlled by the `RichText` component
- the caret is at the beginning of the block

### Technical considerations

I believe this could get also promoted to stable API with the name `onEnterAtStart` together with `onEnterAtEnd` as it is something that proved to be useful with `__unstableOnSplitAtEnd`.

Optionally, we might also consider making both `onEnterAtStart` and `onEnterAtEnd` more abstracted in `RichText` from `@wordpress/block-editor` and allow passing `true`, example:

```jsx
<RichText onEnterAtStart onEnterAtEnd />
```

that would default to the callback that automatically finds the current client id of the block and inserts a new default block before/after the block.

## Testing Instructions





1. Open the site editor.
2. Add a new empty template or remove all blocks from the currently edited template.
3. Add one of the listed blocks, e.g. Code block.
4. Type something, so it contains any content.
5. Move the caret to the start of the block and press `enter`.
6. New default block should get prepended.
